lib/src/main/java/com/diffplug/spotless/protobuf/BufStep.java

```diff
@@ -64,34 +63,33 @@ public FormatterStep create() {
 
 	private State createState() {
 		String instructions = "https://docs.buf.build/installation";
-		ForeignExe exeAbsPath = ForeignExe.nameAndVersion("buf", version)
+		ForeignExe exe = ForeignExe.nameAndVersion("buf", version)
 				.pathToExe(pathToExe)
 				.versionRegex(Pattern.compile("(\\S*)"))
 				.fixCantFind("Try following the instructions at " + instructions + ", or else tell Spotless where it is with {@code buf().pathToExe('path/to/executable')}");
-		return new State(this, exeAbsPath);
+		return new State(this, exe);
 	}
 
 	@SuppressFBWarnings("SE_TRANSIENT_FIELD_NOT_RESTORED")
 	static class State implements Serializable {
 		private static final long serialVersionUID = -1825662356883926318L;
 		// used for up-to-date checks and caching
 		final String version;
-		final transient ForeignExe exe;
+
 		// used for executing
-		private transient @Nullable List<String> args;
+		private final transient ForeignExe exe;
+		private transient String exeAbsPath;
 
-		State(BufStep step, ForeignExe exeAbsPath) {
+		State(BufStep step, ForeignExe exe) {
 			this.version = step.version;
-			this.exe = Objects.requireNonNull(exeAbsPath);
+			this.exe = Objects.requireNonNull(exe);
 		}
 
 		String format(ProcessRunner runner, String input, File file) throws IOException, InterruptedException {
-			if (args == null) {
-				args = Arrays.asList(
-						exe.confirmVersionAndGetAbsolutePath(),
-						"format",
-						file.getAbsolutePath());
+			if (exeAbsPath == null) {
+				exeAbsPath = exe.confirmVersionAndGetAbsolutePath();
 			}
+			List<String> args = List.of(exeAbsPath, "format", file.getAbsolutePath());
```

Comment:
What about just `args.set(2, file.getAbsolutePath()`? I think `exe.confirmVersionAndGetAbsolutePath();` is fairly expensive, and I don't think we need to call it for every invocation.

Reply:
I refactored to only call `confirmVersionAndGetAbsolutePath` once - did you see that?

Reply:
Good call, looks nice. Sorry for slow turnaround.